Adds new features

JhonathanLepidus opened the pull request:
This PR adds the following features:

- Annotations page, where all submissions with annotations are listed
- A new button with the number of annotations next to the PDF button in the preprint page. When this is clicked, the PDF is opened with the annotations tab opening automatically.
- The same button to the preprint item in the list of preprints

These features were discussed in the [issue](https://github.com/asmecher/hypothesis/issues/22)

JhonathanLepidus committed:
WIP: Get galley annotation number through Hypothesis API

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Vitoria <vitoria@lepidus.com.br>

JhonathanLepidus committed:
Finishes HypothesisClient class

It can now request the number of annotations of a galley to the Hypothesis API

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Vitoria <vitoria@lepidus.com.br>

JhonathanLepidus committed:
Prepares the including of custom template in preprint landing page

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Vitoria <vitoria@lepidus.com.br>

JhonathanLepidus committed:
Creates the template to add the annotation viewer to landing page

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Vitoria <vitoria@lepidus.com.br>

JhonathanLepidus committed:
Removes left margin of annotation viewer

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Switch annotation viewer addin strategy

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Removes commented code

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Brings Hypothesis API call to the template

Also changes some visual styling

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Adds Hypothesis config template

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Uses load event for configuring Hypothesis client

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Hypothesis sidebar opens when clicking in annotation number

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Redesigns annotation viewer to use the same link of the galley

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Creates HypothesisHandler class

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Adds page to list submissions with annotations

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Fixes bug occurred when galleys did not have file

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Requests Hypothesis API for annotations by submissions chunks

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Fixes problem with submissions groups annotations requests

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Get submission object for submissions with annotations

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Stores submissions with annotations in cache

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Set annotations cache to updated through scheduled task

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Fixes bug when adding annotation number viewer

Also changes the path for annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Creates page for listing of submissions with annotations

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Sets locales for annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Fix bug occurred for submissions without publications

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Handler returns submissions' annotations

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Sets annotations retrieving from cache

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

richardauzier-afk committed:
Set up the display of annotations

Signed-off-by: Richard <richard@lepidus.com.br>
Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Adds styling for annotations

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Adds Annotation class

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Retrieves annotation data from API response

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Exhibits new annotation info in annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
WIP: Read more for annotation targets

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Adds read more button for annotation targets

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Adds read more button for annotation content

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by: Richard <richard@lepidus.com.br>

JhonathanLepidus committed:
Adds submission metadata to annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>
Signed-off-by> Richard <richard@lepidus.com.br>

lucasDevLepidus committed:
Remove search box

Issue: documentacao-e-tarefas/scielo#396

Signed-off-by: Lucas Anselmo <lucasanselmo@lepidus.com.br>
Signed-off-by: Laís Reis <lais@lepidus.com.br>

JhonathanLepidus committed:
Adds order by date of last annotation

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Adds ordering by date published

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Adds select to order of submissions in annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
OrderBy is selected according to page's URL

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Changes annotations page ordering when one is selected

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Keeps selected order through annotations pagination

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Refactors code of annotations ordering functions

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Changes default ordering to order by last annotation

Signed-off-by: Jhon <jhon@lepidus.com.br>

asmecher commented on templates/submissionAnnotations.tpl:
```diff
@@ -0,0 +1,66 @@
+			</div>
+			{/if}
+			<span class="annotation_content">
+				<blockquote>{$annotation->content}</blockquote>
```
I'm not sure whether annotations should be allowed to contain limited HTML, but this should probably receive at least limited escaping for XSS attacks -- and probably also the other annotation attributes (`user` and `target`) too.

JhonathanLepidus replied:
[Changes done](https://github.com/asmecher/hypothesis/pull/23/commits/b7f93283bf7a22a7faa85a7fcf3b96b6668548d4)!

JhonathanLepidus committed:
Escapes annotations content in annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>

asmecher commented:
Thanks, @JhonathanLepidus! @defstat, would you mind taking a deeper look?

defstat commented:
@JhonathanLepidus two questions on this:

I can see that the description of the issue defines  

> A new button with the number of annotations next to the PDF button in the preprint page.

1. I don't seem to see this button - should I do something in order to see it?
2. Does that work for OPS (because you are referring to preprints) or it should work across applications, namely OJS, OMP and OPS?

JhonathanLepidus commented:
@defstat sorry for missing this information. In fact, there should be an integration with the theme plugin in order for it to work. The integration is the calling of the hook `Hypothesis::annotationNumber` in the place where it is intended to display the button.

For example, we did this implementation in the SciELO Preprints theme: [example](https://github.com/lepidus/scieloTheme/blob/be217f74fa0fd334827f5c03c85cf3bb1cb5ba11/templates/frontend/objects/preprint_details.tpl#L282). It should work in other applications besides OPS.

We did it this way because it was simpler for us, but I think it can be a problem for others to use. Do you think we should change it?

JhonathanLepidus commented:
We were able to identify an alternative implementation. We can change the plugin so that it uses template filters to add the button with the number of annotations next to the galley buttons.

This way, this feature could work on both platforms (OJS/OPS) without needing to be linked to a specific theme.

defstat commented:
Hi @JhonathanLepidus. Thanks for that - Yes, the plugins should be "self contained", so your initial approach should be managed somehow. @jardakotesovec is a better person to talk about front end implementations, so I leave that for him. 

Do you plan to migrate that change to `stable-3.4` and/or `main` branch?

Also could you provide some screenshots on the intended end-result of the features this implementation adds?

JhonathanLepidus commented:
Hi! Yes, we plan to make these features available for the version compatible with stable 3.4.0 as well. Below are some screenshots of these features.

### The annotations page:
![annotations_page](https://github.com/asmecher/hypothesis/assets/60978081/896ea7cb-0428-4ea1-a4a6-d636bd1f7128)
---

### Button with annotations number on archive page
![archive_page](https://github.com/asmecher/hypothesis/assets/60978081/e750249b-c515-44de-97e0-22b57759ddd0)
---

### Button with annotations number on preprint page
![preprint_page](https://github.com/asmecher/hypothesis/assets/60978081/b866484b-53dc-4581-84bd-f9bc56744714)

JhonathanLepidus committed:
Makes plugin fully compatible with OJS and OPS

Also removes dependency for others plugins

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Minor fixes in annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus commented:
Hi @defstat

We have updated the code in the source branch of this pull request, in order to bring the following improvements:

- Updating of the cache cointaining the submissions with annotations through a scheduled task by the Acron plugin
- The buttons with the number of annotations are now added without the use of a dependency with the theme plugin
- Minor adjustements that make the new features fully compatible with both OJS and OPS

The features have also been adapted to the version compatible with OJS/OPS 3.4.0 and are ready to be shipped at a later moment.

defstat commented:
Hi @JhonathanLepidus! Thanks for all that work! It looks great, and I see that you have adapted the newer/more fluent syntax regarding 

Only one question about the caching - is the caching mechanism going to affect the realtime update of the annotation display?

If you have PRs for 3.4 version, please don't hesitate to add those in https://github.com/asmecher/hypothesis/issues/22 so that they can be reviewed also.

JhonathanLepidus commented:
Hi @defstat , thanks for the feedback.

The caching mechanism does not affect the display of annotations when viewing the PDF. The cached items are only used for display on the annotations page and do not affect other parts of the application.

We will open a pull request with our contributions to 3.4.0.

Edit: [PR opened](https://github.com/asmecher/hypothesis/pull/24)

defstat commented:
@JhonathanLepidus thanks for all this work. Can you share the state of this change?

JhonathanLepidus commented:
HI @defstat !

The changes are ready to be merged. They're present in this pull request (for 3.3.0 version) and in [this other one](https://github.com/asmecher/hypothesis/pull/24) (for 3.4.0 version).

If you have any other doubt, please contact me.

defstat commented on HypothesisPlugin.inc.php:
```diff
@@ -76,6 +84,83 @@ function changePdfjsPath($output, $templateMgr) {
+		];
+
+		if (in_array($template, $pagesToInsert)) {
+            $request = Application::get()->getRequest();
```
Here it seems that we have a different Indentation (spaces) than in other places (tabs) which is the general approach for v3.3. 

This is true also elsewhere in the plugin's code

defstat commented on HypothesisPlugin.inc.php:
```diff
@@ -76,6 +84,83 @@ function changePdfjsPath($output, $templateMgr) {
+    {
+        $request = Application::get()->getRequest();
+        $templateMgr = TemplateManager::getManager($request);
+        $handlerUrl = $request->getDispatcher()->url($request, ROUTE_COMPONENT, null, 'plugins.generic.hypothesis.controllers.HypothesisHandler');
```
Are we sure that the scheduled task is executed successfully? I am getting a 

```
PHP Fatal error:  Uncaught Error: Call to a member function url() on null in ojs-3.3.0-17\plugins\generic\hypothesis\HypothesisPlugin.inc.php:158
```

When trying to execute `php runScheduledTasks.php -h`

defstat reviewed (changes requested):
Thanks @JhonathanLepidus so much for those changes, and I am sorry for delaying this. Just one or two comments here:

1. I have tried to annotate something on a PDF and then to check that annotation on the interface, but it did not seem to be visible. Maybe I am doing something wrong, but if you can, please try to test the plugin's full functionality on stable-3_3_0 branch, with PHP 7.3 which is the oldest we support on stable-3_3_0.
2. Please try the same on stable-3_4_0 for the other PR you are mentioning, and PHP 8.0 which is the oldest PHP version we support on that stable branch.

JhonathanLepidus committed:
Fixes error in task for updating annotations cache

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Indents code using tabs instead of spaces

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Fixes minor leftover error in indentation

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Fixes error when adding javascript to frontend

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus commented:
Hi @defstat , thanks for the review of these changes.

I have tested both versions of these changes in PHP 7.3 and PHP 8.0. Some fixes were needed for the first, but the second worked well.

I also fixed one bug in the 3.3.0-compatible version regarding the scheduled task and solved the problem with the `runScheduledTasks` tool.

I formatted the PHP code using tabs in the 3.3.0 version. However, I noticed the plugin doesn't follow the PSR-12 standard in 3.4.0, so I didn't know what to do in this case.

defstat commented:
Hi @JhonathanLepidus, because I have some functionality questions regarding this, could we set-up a quick demo meeting for the new features?

JhonathanLepidus commented:
Hi @defstat . Sure, I'm available for a meeting on any day of the week, except on wednesdays afternoons (considering brazilian time).

defstat commented:
Thanks @JhonathanLepidus! My email is defstat@gmail.com. Send me when ever you can an email so that we can schedule something there.

defstat commented:
Hi @JhonathanLepidus! do we have any development on this?

JhonathanLepidus committed:
Fixes bugs with retrieving of submissions annotations

The bugs were regarding the retrieving of a group of annotations and submissions which used custom url paths

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus commented:
Hi @defstat ! These changes were not firstly prioritized, but I found some time to do these tests. As we discussed, I did the tests with the PKP datasets. I found two bugs on the 3.3.0 and 3.4.0 changes.

When retrieving the annotations, the submissions were divided in groups and the Hypothesis API was queried for each group (to reduce time in large databases). The bug affected the last group queried. Since the PKP dataset has only two submissions, there was only one group, which was not queried.

The second was regarding submissions with a custom URL path. This bug was present only on the 3.3.0 changes.

I still did not implement the new item for the site's navigation menu. Please, let me know if you still have problems testing these features.

asmecher commented:
@defstat, I think the next step on this one is with you.

JhonathanLepidus committed:
Adds NMI type for annotations page

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Adds navigation menu item when plugin is enabled

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus committed:
Renames locale keys for navigation menu item

Signed-off-by: Jhon <jhon@lepidus.com.br>

JhonathanLepidus commented:
Hi @defstat

As we agreed previously, I added a new navigation menu item to ease the access of users to the annotations page. This feature is now present in both this PR and the [PR for 3.4.0](https://github.com/asmecher/hypothesis/pull/24).

<img width="800" height="344" alt="annotations_nmi" src="https://github.com/user-attachments/assets/1b93c739-f4d5-4808-9f70-3a4938470979" />
